source/common/stats/symbol_table_impl.h

```diff
@@ -19,36 +23,166 @@
 namespace Envoy {
 namespace Stats {
 
+/** A Symbol represents a string-token with a small index. */
 using Symbol = uint32_t;
+
+/** Efficient byte-encoded storage an array of tokens, which are typically < 127 */
+using SymbolStorage = uint8_t[];
+
+/** Transient representations of a vector of 32-bit symbols */
 using SymbolVec = std::vector<Symbol>;
 
+class StatName;
+
+/**
+ * Represents an 8-bit encoding of a vector of symbols, used as a transient
+ * representation during encoding and prior to retained allocation.
+ */
+class SymbolEncoding {
+public:
+  /**
+   * Before destructing SymbolEncoding, you must call moveToStorage. This
+   * transfers ownership, and in particular, the responsibility to call
+   * SymbolTable::clear() on all referenced symbols. If we ever wanted
+   * to be able to destruct a SymbolEncoding without transferring it
+   * we could add a clear(SymbolTable&) method.
+   */
+  ~SymbolEncoding();
+
+  /**
+   * Encodes a token into the vec.
+   *
+   * @param symbol the symbol to encode.
+   */
+  void addSymbol(Symbol symbol);
+
+  /**
+   * Decodes a uint8_t array into a SymbolVec.
+   */
+  static SymbolVec decodeSymbols(const SymbolStorage array, size_t size);
```

Comment:
nit: generally prefer explicit types. I would probably replace size_t everywhere with either uint32_t or uint64_t

Reply:
Sure that's fine.